pkg/ccl/backupccl/backup_compaction.go

```diff
@@ -0,0 +1,497 @@
+		if err := backupinfo.WriteMetadataWithExternalSSTs(ctx, store, encryption,
+			kmsEnv, backupManifest); err != nil {
+			return err
+		}
+	}
+
+	statsTable := getTableStatsForBackup(ctx, p.ExecCfg().TableStatsCache, backupManifest.Descriptors)
+	if err := backupinfo.WriteTableStatistics(ctx, store, encryption, kmsEnv, &statsTable); err != nil {
+		return err
+	}
+
+	if backupinfo.WriteMetadataSST.Get(&p.ExecCfg().Settings.SV) {
+		if err := backupinfo.WriteBackupMetadataSST(ctx, store, encryption, kmsEnv, backupManifest,
+			statsTable.Statistics); err != nil {
+			err = errors.Wrap(err, "writing forward-compat metadata sst")
+			if !build.IsRelease() {
+				return err
+			}
+			log.Warningf(ctx, "%+v", err)
+		}
+	}
+	return nil
+}
+
+func runCompaction(
+	ctx context.Context,
+	p sql.JobExecContext,
+	encryption *jobspb.BackupEncryptionOptions,
+	entries chan execinfrapb.RestoreSpanEntry,
+	details jobspb.BackupDetails,
+	manifest *backuppb.BackupManifest,
+	progCh chan execinfrapb.RemoteProducerMetadata_BulkProcessorProgress,
+	store cloud.ExternalStorage,
+) error {
+	defer close(progCh)
+	var encryptionOptions *kvpb.FileEncryptionOptions
+	if encryption != nil {
+		encryptionOptions = &kvpb.FileEncryptionOptions{Key: encryption.Key}
+	}
+	sinkConf := sstSinkConf{
+		id:       p.ExecCfg().DistSQLSrv.NodeID.SQLInstanceID(),
+		enc:      encryptionOptions,
+		progCh:   progCh,
+		settings: &p.ExecCfg().Settings.SV,
+	}
+	sink := makeFileSSTSink(sinkConf, store, nil)
+	sink.elideMode = manifest.ElidedPrefix
+	defer func() {
+		if err := sink.flush(ctx); err != nil {
+			log.Warningf(ctx, "failed to flush sink: %v", err)
+		}
+		logClose(ctx, sink, "SST sink")
+	}()
+	for {
+		select {
+		case entry, ok := <-entries:
+			if !ok {
+				return nil
+			}
+
+			sstIter, err := openSSTs(ctx, p, entry, encryption, details)
+			if err != nil {
+				return errors.Wrap(err, "opening SSTs")
+			}
+
+			if err := processSpanEntry(ctx, sstIter, sink, manifest); err != nil {
+				return errors.Wrap(err, "processing span entry")
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func processSpanEntry(
+	ctx context.Context, sstIter mergedSST, sink *fileSSTSink, manifest *backuppb.BackupManifest,
+) error {
+	defer sstIter.cleanup()
+	entry := sstIter.entry
+	startKeyMVCC, endKeyMVCC := storage.MVCCKey{Key: entry.Span.Key}, storage.MVCCKey{Key: entry.Span.EndKey}
+	prefix, err := elidedPrefix(entry.Span.Key, entry.ElidedPrefix)
+	if err != nil {
+		return err
+	}
+	if prefix != nil {
+		startKeyMVCC.Key = bytes.TrimPrefix(startKeyMVCC.Key, prefix)
+		endKeyMVCC.Key = bytes.TrimPrefix(endKeyMVCC.Key, prefix)
+	}
+	iter := sstIter.iter
+	for iter.SeekGE(startKeyMVCC); ; iter.NextKey() {
+		ok, err := iter.Valid()
+		if err != nil {
+			return err
+		}
+		if !ok {
+			break
+		}
+		key := iter.UnsafeKey()
+		if !key.Less(endKeyMVCC) {
+			break
+		}
+		uv, err := iter.UnsafeValue()
+		v, err := storage.DecodeMVCCValueAndErr(uv, err)
```

Comment:
nit: no need to decode the value. save some compute by passing the raw value to the writer and then call `PutRawMVCC`